src/sema/diagnostics.rs

```diff
@@ -2,51 +2,119 @@ use super::ast::{Diagnostic, Level, Namespace};
 use crate::file_resolver::FileResolver;
 use crate::parser::pt::Loc;
 use codespan_reporting::{diagnostic, files, term};
+use itertools::Itertools;
 use serde::Serialize;
+use std::slice::Iter;
 use std::{io, sync::Arc};
 
-/// Print the diagnostics to stderr with fancy formatting
-pub fn print_diagnostics(cache: &FileResolver, ns: &Namespace, debug: bool) {
-    let (files, file_id) = convert_files(ns, cache);
+#[derive(Default)]
+pub struct Diagnostics {
+    contents: Vec<Diagnostic>,
+    has_error: bool,
+}
 
-    let writer = term::termcolor::StandardStream::stderr(term::termcolor::ColorChoice::Always);
-    let config = term::Config::default();
+impl Diagnostics {
+    pub fn any_errors(&self) -> bool {
+        self.has_error
+    }
 
-    for msg in &ns.diagnostics {
-        if msg.level == Level::Debug && !debug {
-            continue;
+    /// Borrow the contents vector
+    /// We should not modify the contents without this data structure
+    pub fn get_contents(&self) -> &Vec<Diagnostic> {
+        &self.contents
+    }
+
+    pub fn iter(&self) -> Iter<Diagnostic> {
```

Comment:
Is this possible with `impl Iterator for Diagnostics`

Reply:
I don't think that is necessary. Implementing the `Iterator` trait for diagnostics would require some extra code to handle functions like `next()`, which keeps track of the last returned element.

Reply:
I think you're right, I can't find a way of doing this with `impl Iterator` either.